src/dc/nipals.js

```diff
@@ -47,16 +40,19 @@ export default class nipals {
       if (Y) {
         q = Y.transpose().mmul(t).div(t.transpose().mmul(t).get(0, 0));
         q = q.div(q.norm());
-
         u = Y.mmul(q).div(q.transpose().mmul(q).get(0, 0));
       } else {
         u = t;
       }
-    }
+      counter++;
+      if (counter === maxIterations) {
+        throw new Error('The maximum number of iterations has been reached');
+      }
+    } while (diff > terminationCriteria);
 
     if (Y) {
       let p = X.transpose().mmul(t).div(t.transpose().mmul(t).get(0, 0));
-      p = p.div(p.norm());
```

Comment:
I would leave the normalisation as before, I think this is wrong in the reference code